File: hashTable/project1/hashSet_w3schools_trial.py
# my own edit
def hash_function(value):
    sum_of_chars = 0
    for char in value:
        sum_of_chars += ord(char)
        # print("This is the character",char)
        # print(ord(char))
        # print(sum_of_chars)
        # print()
    return sum_of_chars % 10

def contains(name):
    index = hash_function(name)
    return my_hash_set[index] == name

# arrays
my_array = ['Pete', 'Jones', 'Lisa', 'Bob', 'Siri']
my_hash_set = [None,None,None,None,None,None,None,None,None,None]

print("'Bob' has hash code:",hash_function('Bob'))
my_hash_set[hash_function('Bob')] = 'Bob'
print("'Lisa' has hash code:",hash_function('Lisa'))
my_hash_set[hash_function('Lisa')] = 'Lisa'
print("'Jones' has hash code:",hash_function('Jones'))
my_hash_set[hash_function('Jones')] = 'Jones'
print("'Siri' has hash code:",hash_function('Siri'))
my_hash_set[hash_function('Siri')] = 'Siri'
print("'Pete' has hash code:",hash_function('Pete'),"\n")
my_hash_set[hash_function('Pete')] = 'Pete'


iteration = 0
for value in my_hash_set:
    print(iteration, '.', value)
    iteration += 1

# Check if 'Pete' is in the Hash Set
print("\n'Pete' is in the Hash Set:",contains('Pete'))


# the proper w3schools code
my_hash_set = [
    [None],
    ['Jones'],
    [None],
    ['Lisa'],
    [None],
    ['Bob'],
    [None],
    ['Siri'],
    ['Pete'],
    [None]
]

def hash_function(value):
    return sum(ord(char) for char in value) % 10
    
def add(value):
    index = hash_function(value)
    bucket = my_hash_set[index]
    if value not in bucket:
        bucket.append(value)
        
def contains(value):
    index = hash_function(value)
    bucket = my_hash_set[index]
    return value in bucket

add('Stuart')

print(my_hash_set)
print('Contains Stuart:',contains('Stuart'))

#Python

File: hashTable/project1/hashSet_w3schools_trial.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUCKET_COUNT 10
#define MAX_STRING_LENGTH 256 // Define a maximum string length

typedef struct Node {
    char value[MAX_STRING_LENGTH]; // Use a fixed-size array for values
    struct Node* next;
} Node;

typedef struct {
    Node* buckets[BUCKET_COUNT];
} SimpleHashSet;

unsigned int hashFunction(const char* value) {
    unsigned int hash = 0;
    while (*value) {
        hash = (hash + *value++) % BUCKET_COUNT;
    }
    return hash;
}

void initHashSet(SimpleHashSet* set) {
    for (int i = 0; i < BUCKET_COUNT; i++) {
        set->buckets[i] = NULL;
    }
}

int contains(SimpleHashSet *set, const char *value) {
    unsigned int index = hashFunction(value);
    Node *node = set->buckets[index];
    while (node != NULL) {
        if (strcmp(node->value, value) == 0) {
            return 1; // Found
        }
        node = node->next;
    }
    return 0; // Not found
}

void removeValue(SimpleHashSet *set, const char *value) {
    unsigned int index = hashFunction(value);
    Node **node = &(set->buckets[index]);
    while (*node != NULL) {
        if (strcmp((*node)->value, value) == 0) {
            Node *toDelete = *node;
            *node = (*node)->next;
            free(toDelete); // Just free the node, not toDelete->value
            return; // Value removed
        }
        node = &((*node)->next);
    }
}

void add(SimpleHashSet* set, const char* value) {
    unsigned int index = hashFunction(value);
    
    // Check if the value already exists to avoid duplicates
    Node* current = set->buckets[index];
    while (current != NULL) {
        if (strcmp(current->value, value) == 0) {
            // Value already exists, do not add again
            return;
        }
        current = current->next;
    }
    
    // Add new value
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (!newNode) return; // Handle malloc failure
    
    strncpy(newNode->value, value, MAX_STRING_LENGTH);
    newNode->value[MAX_STRING_LENGTH - 1] = '\0'; // Ensure null-termination
    newNode->next = set->buckets[index];
    set->buckets[index] = newNode;
}

void printHashSet(SimpleHashSet *set) {
    printf("Hash Set Contents:\n");
    for (int i = 0; i < BUCKET_COUNT; i++) {
        Node *node = set->buckets[i];
        printf("Bucket %d: ", i);
        while (node) {
            printf("%s ", node->value);
            node = node->next;
        }
        printf("\n");
    }
}

int main() {
    SimpleHashSet hashSet;
    initHashSet(&hashSet);

    add(&hashSet, "Charlotte");
    add(&hashSet, "Thomas");
    add(&hashSet, "Jens");
    add(&hashSet, "Peter");
    add(&hashSet, "Lisa");
    add(&hashSet, "Adele");
    add(&hashSet, "Michaela");
    add(&hashSet, "Bob");

    printHashSet(&hashSet);

    printf("\n'Peter' is in the set: %s\n", contains(&hashSet, "Peter") ? "true" : "false");
    printf("Removing 'Peter'\n");
    removeValue(&hashSet, "Peter");
    printf("'Peter' is in the set: %s\n", contains(&hashSet, "Peter") ? "true" : "false");
    printf("'Adele' has hash code: %u\n", hashFunction("Adele"));

    // Free memory, omitted for brevity
    return 0;
}

// C